### PR TITLE
Fix invalid memory access during reduction whose output is large to exceed 32-bit

### LINKED ISSUE #1766: Illegal memory access in `cp.sum`
I'm encountering an error where CuPy fails to de-allocate GPU memory after processing a sum operation after a reshape operation.  However, this only occurs if a certain memory threshold is passed.  In the code example, I start with two arrays, reshape them, and then perform the cp.sum operation.  If n is set to <800, the error does not occur, but if n>=1000 the errors occurs (up to failure to allocate due to arrays being too large).  A simple matrix multiplication on the starting 'a' and 'b' to get 'c' does not produce this error.

```
import cupy as cp

n = 1000
a = cp.ones([n,256,100], dtype=cp.float32)
b = cp.ones([n,100,100], dtype=cp.float32)

a = cp.reshape(a, [n,256,100,1])
b = cp.reshape(b, [n,1,100,100])
c = cp.sum(a*b, axis=-2)

print('Complete.')
```

When the code is run and the error occurs, the traceback is as follows.
```
Complete.
Traceback (most recent call last):
  File "cupy/cuda/driver.pyx", line 192, in cupy.cuda.driver.moduleUnload
  File "cupy/cuda/driver.pyx", line 81, in cupy.cuda.driver.check_status
cupy.cuda.driver.CUDADriverError: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
Exception ignored in: 'cupy.cuda.function.Module.__dealloc__'
Traceback (most recent call last):
  File "cupy/cuda/driver.pyx", line 192, in cupy.cuda.driver.moduleUnload
  File "cupy/cuda/driver.pyx", line 81, in cupy.cuda.driver.check_status
cupy.cuda.driver.CUDADriverError: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
```

I have tested this in CuPy 4.5.0 and CuPy 5.0.0.  I am using NVIDIA Driver Version 390.48 and CUDA 9.0, with the appropriate Cupy installation.

### PR DESCRIPTION
**Pull request by kmaehashi**
Fix #1766.

* `_J_offset` can be up to `2^9` and `_out_ind.size.size()` (size_t) can be up to `2^64` (capped at max GPU memory).  Thus `_j_offset` should be typed as  `long long`.
* `gridDim.x` (unsigned int) can be up to `2^31` and `_block_stride` (int) can be up to `2^9`. Thus `_i_base` and `_i` should be typed as `long long`.

### CODE REVIEW

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
This PR is ready for review.
Jenkins, test this please.

File: cupy/core/reduction.pxi [hunk 1]
```diff
@@ -32,15 +32,16 @@ extern "C" __global__ void ${name}(${params}) {
   unsigned int _tid = threadIdx.x;
 
   int _J_offset = _tid >> __popc(_block_stride - 1);  // _tid / _block_stride
-  int _j_offset = _J_offset * _out_ind.size();
+  long long _j_offset = (long long)_J_offset * _out_ind.size();
```
Comment (asi1024) [inline cupy/core/reduction.pxi]: `ptrdiff_t` might be better than `long long` from the point of view of consistency.
https://github.com/kmaehashi/cupy/blob/fix-memory-corruption/cupy/core/include/cupy/carray.cuh#L167

Comment (kmaehashi) [reply cupy/core/reduction.pxi]: Agree with that, fixed.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**